helm-app-operator/pkg/helm/controller/reconcile.go

```diff
@@ -71,52 +76,87 @@ func (r HelmOperatorReconciler) Reconcile(request reconcile.Request) (reconcile.
+		}
 		finalizers := []string{}
 		for _, pendingFinalizer := range pendingFinalizers {
 			if pendingFinalizer != finalizer {
 				finalizers = append(finalizers, pendingFinalizer)
 			}
 		}
 		o.SetFinalizers(finalizers)
-		err := r.Client.Update(context.TODO(), o)
+		err = r.Client.Update(context.TODO(), o)
 		return reconcile.Result{}, err
 	}
 
-	updatedRelease, needsUpdate, err := manager.ReconcileRelease()
-	if err != nil {
-		logrus.Errorf("failed to reconcile release for %s: %s", util.ResourceString(o), err)
-		return reconcile.Result{}, err
-	}
+	if !manager.IsInstalled() {
+		installedRelease, err := manager.InstallRelease(context.TODO())
+		if err != nil {
+			logrus.Errorf("failed to install release for %s release=%s: %s", util.ResourceString(o), releaseName, err)
+			return reconcile.Result{}, err
+		}
+		diff := util.Diff("", installedRelease.GetManifest())
+		logrus.Infof("Installed release for %s release=%s; diff:\n%s", util.ResourceString(o), releaseName, diff)
 
-	if needsUpdate {
-		status := types.StatusFor(o)
-		status.SetRelease(updatedRelease)
-		// TODO(alecmerdler): Call `status.SetPhase()` with `NOTES.txt` of rendered Chart
-		status.SetPhase(types.PhaseApplied, types.ReasonApplySuccessful, "")
-		o.Object["status"] = status
+		status.SetRelease(installedRelease)
+		status.SetPhase(types.PhaseApplied, types.ReasonApplySuccessful, installedRelease.GetInfo().GetStatus().GetNotes())
+		err = r.updateResource(o, status)
+		return reconcile.Result{RequeueAfter: r.ResyncPeriod}, err
+	}
 
-		err = r.Client.Update(context.TODO(), o)
+	if manager.IsUpdateRequired() {
+		previousRelease, updatedRelease, err := manager.UpdateRelease(context.TODO())
 		if err != nil {
-			logrus.Errorf("failed to update resource status for %s: %s", util.ResourceString(o), err)
+			logrus.Errorf("failed to update release for %s release=%s: %s", util.ResourceString(o), releaseName, err)
 			return reconcile.Result{}, err
 		}
+		diff := util.Diff(previousRelease.GetManifest(), updatedRelease.GetManifest())
+		logrus.Infof("Updated release for %s release=%s; diff:\n%s", util.ResourceString(o), releaseName, diff)
+
+		status.SetRelease(updatedRelease)
+		status.SetPhase(types.PhaseApplied, types.ReasonApplySuccessful, updatedRelease.GetInfo().GetStatus().GetNotes())
+		err = r.updateResource(o, status)
+		return reconcile.Result{RequeueAfter: r.ResyncPeriod}, err
 	}
 
+	_, err = manager.ReconcileRelease(context.TODO())
+	if err != nil {
+		logrus.Errorf("failed to reconcile release for %s release=%s: %s", util.ResourceString(o), releaseName, err)
+		return reconcile.Result{}, err
+	}
+	logrus.Infof("Reconciled release for %s release=%s", util.ResourceString(o), releaseName)
+
 	return reconcile.Result{RequeueAfter: r.ResyncPeriod}, nil
 }
 
+func (r HelmOperatorReconciler) updateResource(o *unstructured.Unstructured, status *types.HelmAppStatus) error {
```

Comment:
Just a nit but how about calling this `updateResourceStatus()` to make it more specific?